### ISSUE TITLE
Match statement syntax error does not obey version guards

### ISSUE THREAD

**Issue by wyattscarpenter** [python/mypy#19923]
Originally discovered by @dmarteau in https://github.com/python/mypy/issues/19892.

I wrote and ran the following two tests under Python 3.12 and they succeeded. 
> 
> ```py
> [case testUnreachableAfterToplevelAssertFlaggy]
> # flags: --python-version=3.9 --show-error-codes
> # This test may false negative succeed if you are actually running python 3.9
> import sys
> reveal_type(0)  # N: Revealed type is "Literal[0]?"
> x: int | tuple[int]  # E: X | Y syntax for unions requires Python 3.10  [syntax]
> "a"+1  # E: Unsupported operand types for + ("str" and "int")  [operator]
> assert sys.version_info > (3,9)
> reveal_type('')  # No error here :-)
> x: int | tuple[int] # a little 3.10 syntax to check
> [builtins fixtures/ops.pyi]
> 
> [case testUnreachableAfterToplevelAssertConfigy]
> # This test may false negative succeed if you are act>
> # flags: --config-file=tmp/foo.ini --show-error-codes
> import sys
> reveal_type(0)  # N: Revealed type is "Literal[0]?"
> x: int | tuple[int]  # E: X | Y syntax for unions requires Python 3.10  [syntax]
> "a"+1  # E: Unsupported operand types for + ("str" and "int")  [operator]
> assert sys.version_info > (3,9)
> reveal_type('')  # No error here :-)
> x: int | tuple[int] # a little 3.10 syntax to check
> 
> [file foo.ini]
> \[mypy]
> python_version=3.9
> 
> [builtins fixtures/ops.pyi]
> ```

**Comment by wyattscarpenter** [comment]
Possible clues: the union syntax error comes from https://github.com/python/mypy/blob/16cd4c5215aa2301b50341c22048ffb8a4737a63/mypy/typeanal.py#L1386, but I can't even figure out where the match syntax error comes from from its string. In semanal.py there is some stuff https://github.com/python/mypy/blob/16cd4c5215aa2301b50341c22048ffb8a4737a63/mypy/semanal.py#L5541 (Maybe we don't handle the syntax error, and it bubbles out of our parser and we report it as a fallback? Just a guess.)

There is historical context that might be of use https://github.com/python/mypy/pull/10191/files

**Comment by sterliakov** [comment]
Yes, exactly: if a parser cannot parse the complete file, we don't know anything about your version guards, we don't even have a tree to look for them. 

* `X | Y` is not a parsing `SyntaxError`, it's just a gentle reminder from us. We do have a full parsed file, we just see something that will raise at runtime (but not a `SyntaxError`). If there's a preceding version guard, this is fine.
* `match` on py3.9 is a hard `SyntaxError` raised by parser, because the file does not contain valid py3.9 code. This error is blocking, and we stop the checking right there - we don't have any tree to actually check.

**Comment by wyattscarpenter** [comment]
Dang! Well, maybe after mypy moves to a new faster parser https://github.com/python/mypy/issues/19776 this will possible.

Thank you for the excellent explanation 🙂

**Comment by wyattscarpenter** [comment]
Presumably this issue can be closed once Python 3.9 support is dropped. https://github.com/python/mypy/issues/20154

**Comment by hauntsaninja** [comment]
Yeah, this is a won't fix for the reasons sterliakov mentions

**Comment by hauntsaninja** [comment]
(#20154 is about removing support for running with Python 3.9, we will probably only remove support for checking Python 3.9 once typeshed does)

**Comment by A5rocks** [comment]
I'm not completely sure about this, because Python 3.12 *should* be able to parse the code. I guess "do we raise an error for version-guarded match statements" would then be slightly inconsistent (based on if the error comes from `ast` or from us), but nothing a little error message improvement couldn't fix.
